src/RA_Core.cpp

```diff
@@ -994,52 +996,49 @@ API void CCONV _RA_SavePreferences()
-    {
-        FileStream fs(pf);
-        Writer<FileStream> writer(fs);
+    std::ofstream ofile{ ra::PrefsFilename(), std::ios::binary };
 
-        Document doc;
-        doc.SetObject();
-
-        Document::AllocatorType& a = doc.GetAllocator();
-        doc.AddMember("Username", StringRef(RAUsers::LocalUser().Username().c_str()), a);
-        doc.AddMember("Token", StringRef(RAUsers::LocalUser().Token().c_str()), a);
-        doc.AddMember("Hardcore Active", g_bHardcoreModeActive, a);
-        doc.AddMember("Leaderboards Active", g_bLeaderboardsActive, a);
-        doc.AddMember("Leaderboard Notification Display", g_bLBDisplayNotification, a);
-        doc.AddMember("Leaderboard Counter Display", g_bLBDisplayCounter, a);
-        doc.AddMember("Leaderboard Scoreboard Display", g_bLBDisplayScoreboard, a);
-        doc.AddMember("Num Background Threads", g_nNumHTTPThreads, a);
-        doc.AddMember("ROM Directory", StringRef(g_sROMDirLocation.c_str()), a);
-
-        Value positions(kObjectType);
-        for (WindowPositionMap::const_iterator iter = g_mWindowPositions.begin(); iter != g_mWindowPositions.end(); ++iter)
-        {
-            Value rect(kObjectType);
-            if (iter->second.nLeft != WindowPosition::nUnset)
-                rect.AddMember("X", iter->second.nLeft, a);
-            if (iter->second.nTop != WindowPosition::nUnset)
-                rect.AddMember("Y", iter->second.nTop, a);
-            if (iter->second.nWidth != WindowPosition::nUnset)
-                rect.AddMember("Width", iter->second.nWidth, a);
-            if (iter->second.nHeight != WindowPosition::nUnset)
-                rect.AddMember("Height", iter->second.nHeight, a);
-
-            if (rect.MemberCount() > 0)
-                positions.AddMember(StringRef(iter->first.c_str()), rect, a);
-        }
+    using namespace rapidjson;
+    OStreamWrapper osw{ ofile };
+    Writer<OStreamWrapper> writer{ osw };
 
-        if (positions.MemberCount() > 0)
-            doc.AddMember("Window Positions", positions.Move(), a);
+    Document doc;
+    doc.SetObject();
 
-        doc.Accept(writer);	//	Save
+    Document::AllocatorType& a = doc.GetAllocator();
+    doc.AddMember("Username", StringRef(RAUsers::LocalUser().Username().c_str()), a);
+    doc.AddMember("Token", StringRef(RAUsers::LocalUser().Token().c_str()), a);
+    doc.AddMember("Hardcore Active", g_bHardcoreModeActive, a);
+    doc.AddMember("Leaderboards Active", g_bLeaderboardsActive, a);
+    doc.AddMember("Leaderboard Notification Display", g_bLBDisplayNotification, a);
+    doc.AddMember("Leaderboard Counter Display", g_bLBDisplayCounter, a);
+    doc.AddMember("Leaderboard Scoreboard Display", g_bLBDisplayScoreboard, a);
+    doc.AddMember("Num Background Threads", g_nNumHTTPThreads, a);
+    doc.AddMember("ROM Directory", StringRef(g_sROMDirLocation.c_str()), a);
 
-        fclose(pf);
+    Value positions(kObjectType);
+
+    for (auto& i : g_mWindowPositions)
+    {
```

Comment:
We should be using ranged fors as much as possible (existed since C++11) if you don't need to advance or reverse the iterators. There's also a VS Extension `for each` which works exactly like the .NET `foreach `but has the same semantics as a ranged for loop. there's also `std::for_each` but not totally sure if you guys would be comfortable using `lambdas`. Of course sometimes it's not possible.